cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApiImpl.java

```diff
@@ -200,20 +200,21 @@ private void internalRebalance(String host, int port, String path, String userTa
                             String userTaskID = response.result().getHeader(CC_REST_API_USER_ID_HEADER);
                             JsonObject json = buffer.toJsonObject();
                             if (json.containsKey(CC_REST_API_ERROR_KEY)) {
+                                String errorMessage = json.getString(CC_REST_API_ERROR_KEY);
                                 // If there was a client side error, check whether it was due to not enough data being available ...
-                                if (json.getString(CC_REST_API_ERROR_KEY).contains("NotEnoughValidWindowsException")) {
+                                if (errorMessage.contains("NotEnoughValidWindowsException")) {
                                     CruiseControlRebalanceResponse ccResponse = new CruiseControlRebalanceResponse(userTaskID, json);
                                     ccResponse.setNotEnoughDataForProposal(true);
                                     result.complete(ccResponse);
                                 // ... or one or more brokers doesn't exist on a add/remove brokers rebalance request
-                                } else if (json.getString(CC_REST_API_ERROR_KEY).contains("IllegalArgumentException") &&
-                                            json.getString(CC_REST_API_ERROR_KEY).contains("does not exist.")) {
+                                } else if (errorMessage.contains("IllegalArgumentException") &&
+                                            errorMessage.contains("does not exist.")) {
                                     result.fail(new IllegalArgumentException("Some/all brokers specified don't exist"));
                                 } else {
                                     // If there was any other kind of error propagate this to the operator
-                                    result.fail(new CruiseControlRestException(
+                                    result.fail(new CruiseControlRestException(new Exception(errorMessage.substring(errorMessage.indexOf(":") + 3, errorMessage.indexOf(":", errorMessage.indexOf("Exception")))),
```

Comment:
Let's forget about grabbing the exact exception name from the Cruise Control stacktrace. You were right, it is messier and more work than it is worth. 

Instead, let's just pass the
(1) `errorMessage` from the json as the`CruiseControlRestException`'s message
(2) `stackTrace` from the json as the`CruiseControlRestException`'s cause

Having something like the following:
```
String errorPrefix =  "Error for request: " + host + ":" + port + path + ". Server returned: ";
String message = errorPrefix + json.getString(CC_REST_API_ERROR_KEY);
Throwable cause = new Throwable(errorPrefix + json.getString(CC_REST_API_STACK_TRACE_KEY));
                                    
result.fail(new CruiseControlRestException(message, cause));
```

Then we can call `exception.getCause()` in the KafkaRebalanceAssemblyOperator in places like here [1]
```
LOGGER.errorCr(reconciliation, "Status updated to [NotReady] due to error: {}", exception.getCause());
```
This way, we would get the cause of the rebalance in form of a stackTrace in the logs on the Cluster Operator as evidence as why the rebalanced request failed.

At the same time, the status of the KafkaRebalance resource would be updated with the errorMessage as you have it now

[1] https://github.com/strimzi/strimzi-kafka-operator/blob/0.34.0/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java#L485

Reply:
HI @kyguy, After applying your suggestion I get this in the status --> 
```yaml
kind: KafkaRebalance
metadata:
  creationTimestamp: "2023-06-06T09:34:33Z"
  generation: 1
  labels:
    strimzi.io/cluster: my-cluster
  name: my-rebalance
  namespace: myproject
  resourceVersion: "1019"
  uid: fd37f7b9-51b7-4038-9495-c617b0b2c805
spec: {}
status:
  conditions:
  - lastTransitionTime: "2023-06-06T09:34:33.698427968Z"
    status: "True"
    type: PendingProposal
  observedGeneration: 1
srawat@srawat-mac cruise-control % kubectl get kafkarebalance my-rebalance -n myproject -o yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  creationTimestamp: "2023-06-06T09:34:33Z"
  generation: 1
  labels:
    strimzi.io/cluster: my-cluster
  name: my-rebalance
  namespace: myproject
  resourceVersion: "1109"
  uid: fd37f7b9-51b7-4038-9495-c617b0b2c805
spec: {}
status:
  conditions:
  - lastTransitionTime: "2023-06-06T09:35:48.883269711Z"
    message: "Error for request: my-cluster-cruise-control.myproject.svc:9090/kafkacruisecontrol/rebalance?json=true&dryrun=true&verbose=true&skip_hard_goal_check=false&rebalance_disk=false.
      Server returned: java.util.concurrent.ExecutionException: com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException:
      [RackAwareGoal] Insufficient number of racks to distribute each replica (Current:
      1, Needed: 3). Add at least 2 racks with brokers. Add at least 2 racks with
      brokers.\n\tat java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)\n\tat
      java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.handler.async.AbstractAsyncRequest.getResponse(AbstractAsyncRequest.java:56)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.handler.AbstractRequest.handle(AbstractRequest.java:37)\n\tat
      com.linkedin.kafka.cruisecontrol.KafkaCruiseControlRequestHandler.handlePost(KafkaCruiseControlRequestHandler.java:156)\n\tat
      com.linkedin.kafka.cruisecontrol.KafkaCruiseControlRequestHandler.doGetOrPost(KafkaCruiseControlRequestHandler.java:74)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.ServletRequestHandler.doPost(ServletRequestHandler.java:46)\n\tat
      javax.servlet.http.HttpServlet.service(HttpServlet.java:707)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:790)\n\tat
      org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)\n\tat
      org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:554)\n\tat
      org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)\n\tat
      org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:560)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.security.CruiseControlSecurityHandler.handle(CruiseControlSecurityHandler.java:22)\n\tat
      org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)\n\tat
      org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)\n\tat
      org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)\n\tat
      org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)\n\tat
      org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)\n\tat
      org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)\n\tat
      org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)\n\tat
      org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)\n\tat
      org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)\n\tat
      org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)\n\tat
      org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)\n\tat
      org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)\n\tat
      org.eclipse.jetty.server.Server.handle(Server.java:516)\n\tat org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)\n\tat
      org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)\n\tat org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)\n\tat
      org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)\n\tat
      org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)\n\tat
      org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)\n\tat org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:555)\n\tat
      org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:410)\n\tat
      org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:164)\n\tat
      org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)\n\tat org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)\n\tat
      org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)\n\tat
      org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)\n\tat
      org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)\n\tat
      org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)\n\tat
      org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)\n\tat
      org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)\n\tat
      org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)\n\tat
      java.base/java.lang.Thread.run(Thread.java:833)\nCaused by: com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException:
      [RackAwareGoal] Insufficient number of racks to distribute each replica (Current:
      1, Needed: 3). Add at least 2 racks with brokers. Add at least 2 racks with
      brokers.\n\tat com.linkedin.kafka.cruisecontrol.analyzer.goals.AbstractGoal.optimize(AbstractGoal.java:130)\n\tat
      com.linkedin.kafka.cruisecontrol.analyzer.GoalOptimizer.optimizations(GoalOptimizer.java:467)\n\tat
      com.linkedin.kafka.cruisecontrol.KafkaCruiseControl.optimizations(KafkaCruiseControl.java:622)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.ProposalsRunnable.workWithClusterModel(ProposalsRunnable.java:114)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.GoalBasedOperationRunnable.computeResult(GoalBasedOperationRunnable.java:161)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.RebalanceRunnable.workWithoutClusterModel(RebalanceRunnable.java:118)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.GoalBasedOperationRunnable.computeResult(GoalBasedOperationRunnable.java:173)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.RebalanceRunnable.getResult(RebalanceRunnable.java:90)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.RebalanceRunnable.getResult(RebalanceRunnable.java:30)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.OperationRunnable.run(OperationRunnable.java:45)\n\tat
      com.linkedin.kafka.cruisecontrol.servlet.handler.async.runnable.GoalBasedOperationRunnable.run(GoalBasedOperationRunnable.java:36)\n\tat
      java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)\n\tat
      java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)\n\tat
      java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)\n\t...
      1 more\n"
    reason: Throwable
    status: "True"
    type: NotReady
  observedGeneration: 1
```

Reply:
Hi @ShubhamRwt, these changes should place the errorMessage in the `message` field of the status, not the `cause` (full stacktrace). If you post the code changes made I could take a closer look, otherwise I can try it locally (it should work, I think we are just missing some changes)

Reply:
Try making the changes like here [1], it seems to do the trick!

[1] https://github.com/kyguy/strimzi-kafka-operator/commit/30524292aafa874e53d17914fadb0cc0f98287a6

Reply:
Hi @kyguy, left comments on your commit